File: Python/dp.py
def dfs(i, j, m, dp):
    if dp[i][j]: return dp[i][j]
    n, p = len(m), len(m[0])
    ans = 1
    for x, y in [(1,0),(-1,0),(0,1),(0,-1)]:
        ni, nj = i + x, j + y
        if 0 <= ni < n and 0 <= nj < p and m[ni][nj] > m[i][j]:
            ans = max(ans, 1 + dfs(ni, nj, m, dp))
    dp[i][j] = ans
    return ans

def longestIncreasingPath(m):
    n, p = len(m), len(m[0])
    dp = [[0]*p for _ in range(n)]
    res = 0
    for i in range(n):
        for j in range(p):
            res = max(res, dfs(i, j, m, dp))
    return res

n, p = map(int, input().split())
matrix = [list(map(int, input().split())) for _ in range(n)]
print(longestIncreasingPath(matrix))

File: Java/dp.java
import java.util.*;

public class Main {
    static int dfs(int i, int j, int[][] m, int[][] dp) {
        if (dp[i][j] != 0) return dp[i][j];
        int n = m.length, p = m[0].length, ans = 1;
        int[][] dirs = {{1,0},{-1,0},{0,1},{0,-1}};
        for (int[] d : dirs) {
            int x = i + d[0], y = j + d[1];
            if (x >= 0 && y >= 0 && x < n && y < p && m[x][y] > m[i][j])
                ans = Math.max(ans, 1 + dfs(x, y, m, dp));
        }
        return dp[i][j] = ans;
    }

    static int longestIncreasingPath(int[][] m) {
        int n = m.length, p = m[0].length, res = 0;
        int[][] dp = new int[n][p];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < p; j++)
                res = Math.max(res, dfs(i, j, m, dp));
        return res;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(), p = sc.nextInt();
        int[][] matrix = new int[n][p];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < p; j++)
                matrix[i][j] = sc.nextInt();
        System.out.println(longestIncreasingPath(matrix));
    }
}
